### ISSUE TITLE
ArrayMap .keys() method could return an Iterator.

### ISSUE THREAD

**Issue by DutchGhost** [cberner/raptorq#9]
Currently the .keys() method on ArrayMap returns a vector.

With minor modifications, it can return an Iterator, saving allocations:

```Rust
pub fn keys<'s>(&'s self) -> impl Iterator<Item = usize> + 's {
    self.elements
        .iter()
        .enumerate()
        .filter_map(move |(i, elem)| match elem {
            Some(_) => Some(i + self.offset),
            None => None,
        })
}
```

The only downside of this approach is that the Iterator borrows self for 's, while the Vec approach doesn't borrow once the end of the method is reached. However all your tests still pass using this Iterator approach.

**Comment by cberner** [comment]
Cool. Want to submit a PR? I'm happy to merge this, if it doesn't reduce performance

**Comment by cberner** [comment]
Fixed by #10
